### DISCUSSION TITLE
Coredump support in could hypervisor?

### DISCUSSION THREAD

**Discussion by zhuangel** [cloud-hypervisor/cloud-hypervisor#5014] [General]
Hi, I am new to cloud hypervisor, I want to know enable core dump in release version is useful, once the VMM process run abnormally and quit, we could diagnose exception context from core dump.

As I know, to support this we need:
1 set core dump mechanism corecttly, such as /proc/sys/kernel/core_pattern, ulimit
2 abort when panic occurs in cloud hypervisor, refer to [config](https://doc.rust-lang.org/book/ch09-01-unrecoverable-errors-with-panic.html#unrecoverable-errors-with-panic)
3 add 'dump' config to guest memory, so that we could set MADV_DONTDUMP to these memory region, to speed up core dump process and decrease core dump file size.

Any suggestions, Thanks.

**Comment by sboeuf** [comment]
The ability to coredump the VM is part of the Rust feature gate `guest_debug`.
If you want a release version of Cloud Hypervisor that supports coredump, you simply need to build it with the following command:
```
cargo build --release --features guest_debug
```
And then use `ch-remote` to trigger the dump:
```
target/release/ch-remote --api-socket=/tmp/ch.sock coredump file:///tmp/ch_coredump
```

**Reply by zhuangel** [reply]
Thanks for the replay.

I mean just application coredump for Cloud Hypervisor itself, such as the process killed by some signal or even segment fault, does not include the guest kernel vmcore.

**Reply by sboeuf** [reply]
Sorry I misread your description.

> 1 set core dump mechanism corecttly, such as /proc/sys/kernel/core_pattern, ulimit

This can be performed by the user/management layer after the VMM has been started.

> 2 abort when panic occurs in cloud hypervisor, refer to [config](https://doc.rust-lang.org/book/ch09-01-unrecoverable-errors-with-panic.html#unrecoverable-errors-with-panic)

Again, this can be performed by the user, modifying the `Cargo.toml` file for this special purpose. I don't think we want to include it by default in our `Cargo.toml`.

> add 'dump' config to guest memory, so that we could set `MADV_DONTDUMP` to these memory region, to speed up core dump process and decrease core dump file size.

Reading from the `madvise(2)` documentation, it seems like `MADV_DONTDUMP` doesn't bring any side effect. We could enable it by default on the guest memory regions, no need for a special option or parameter.

**Reply by zhuangel** [reply]
Thanks @sboeuf  for your kindly reply.

For question 1 and 2, I agree with you. 

For question 3, do you think we could set it by default, such as in `create_userspace_mapping`, I found qemu has some flags to control this [https://patchwork.ozlabs.org/project/qemu-devel/patch/201208021944.q72JiGm2018091@int-mx09.intmail.prod.int.phx2.redhat.com/](url)

Of course we could do it in a simple way as you said. Would you mind I create PR for this revise, thanks.

**Reply by sboeuf** [reply]
Well! Looks like @rbradford was too fast :)
He submitted https://github.com/cloud-hypervisor/cloud-hypervisor/pull/5016 to handle point 3.

**Reply by zhuangel** [reply]
That's great, thanks a lot!
